pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -659,23 +659,51 @@ func endpointPortFromServicePort(servicePort *v1.ServicePort, portNum int) *v1.E
 	return epp
 }
 
-// overCapacity returns true if there are more addresses in the provided subsets
-// than the maxCapacity.
-func overCapacity(subsets []v1.EndpointSubset) bool {
+// capacityAnnotationSetCorrectly returns false if number of endpoints is greater than maxCapacity or
+// returns true if underCapacity and the annotation is not set.
+func capacityAnnotationSetCorrectly(annotations map[string]string, subsets []v1.EndpointSubset) bool {
 	numEndpoints := 0
 	for _, subset := range subsets {
 		numEndpoints += len(subset.Addresses) + len(subset.NotReadyAddresses)
 	}
-	return numEndpoints > maxCapacity
+	if numEndpoints > maxCapacity {
+		// If subsets are over capacity, they must be truncated so consider
+		// the annotation as not set correctly
+		return false
+	}
+	_, ok := annotations[v1.EndpointsOverCapacity]
+	return !ok
 }
 
-// capacityAnnotationSetCorrectly returns true if overCapacity() is true and the
-// EndpointsOverCapacity annotation is set to "warning" or if overCapacity()
-// is false and the annotation is not set.
-func capacityAnnotationSetCorrectly(annotations map[string]string, subsets []v1.EndpointSubset) bool {
-	val, ok := annotations[v1.EndpointsOverCapacity]
-	if overCapacity(subsets) {
-		return ok && val == "warning"
+// truncateEndpoints ensures that the number of Addresses does not exceed the maxCapacity. If greater
+// than maxCapacity, the addresses will be truncated so that the number of addresses is maxCapacity.
+func truncateEndpoints(endpoints *v1.Endpoints) bool {
+	canBeAdded := maxCapacity
+
+	for i := range endpoints.Subsets {
+		subset := endpoints.Subsets[i]
+		toBeAdded := len(subset.Addresses) + len(subset.NotReadyAddresses)
+		if canBeAdded-toBeAdded >= 0 {
+			canBeAdded -= toBeAdded
+			continue
+		}
+
+		subset.Addresses = truncateAddresses(subset.Addresses, canBeAdded)
+		canBeAdded -= len(subset.Addresses)
+		subset.NotReadyAddresses = truncateAddresses(subset.NotReadyAddresses, canBeAdded)
```

Comment:
As a follow up to the comment above, I think NotReadyAddresses should have the lowest priority here. We should make sure that all of the ready addresses are accounted for before adding any that aren't ready.

Reply:
That's a good point. Initially I was thinking if we did that it could mean that the addresses will that make the subset will change every time. However as addresses become ready, the same thing occurs so probably not a concern to be worried about.